### PR TITLE
an alternative to pure services

### PR DESCRIPTION
**Pull request by tansaku**
@zsid @hvenables - this is a proposed refactoring related to the area you were working on.

@marianmosley @ZmagoD @decareano @johnnymo87 @mtc2013 would be very interested in your input

This is kind of rough, but I wanted to show the principle of what I meant - relating to Eric Evan's "Domain Driven Design":

* http://blog.fedecarg.com/2009/03/15/domain-driven-design-the-repository/
* https://en.wikipedia.org/wiki/Domain-driven_design

more notes in https://www.pivotaltracker.com/story/show/108935166

### CODE REVIEW

File: app/respositories/proposed_organisation_edit_repository.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class ProposedOrganisationEditRepository
+
+  def self.create(listener, params)
+    unless params[:editor].siteadmin?
+      merge_in_non_published_fields params[:organisation], params
+      send_email_to_superadmin_about_org_edit params[:organisation]
+      listener.flash[:notice] = "Edit is pending admin approval."
+    end
+    ProposedOrganisationEdit.create(params)
```
Comment (johnnymo87) [inline app/respositories/proposed_organisation_edit_repository.rb]: This might fail. How should we handle failure?

**Comment by johnnymo87** [comment]
This looks good!

If repositories are a new domain concept that you want to introduce, my preference is to express this via a namespace ... i.e. `app/repositories/*` and `Repositories::ProposedOrganisationEdit`.

I read through the article you linked on DDD repositories. I see you've titled this PR "An alternative to pure services". Can you expand on what you mean by this? To me, this code looks like a service that's been written with as much `ubiquitous language` as possible. Is there more to distinguish?

File: app/respositories/proposed_organisation_edit_repository.rb (hunk 1 above)
Comment (tansaku) [inline app/respositories/proposed_organisation_edit_repository.rb]: probably in same way we do in services?

**Comment by tansaku** [comment]
when I say alternative to pure services, I just mean that we expand beyond services which can only be accessed by call or run.  The repository has a slightly different vocabulary that ties it to the resource model better I think ...

**Comment by tansaku** [comment]
aha @marianmosley @ZmagoD @decareano @johnnymo87 @mtc2013  and I see others have been talking about repositories in Rails:

* https://blog.lelonek.me/why-is-your-rails-application-still-coupled-to-activerecord-efe34d657c91#.4uxt93r2s
* http://victorsavkin.com/post/41016739721/building-rich-domain-models-in-rails-separating

and the repository pattern in general:

* https://blog.8thlight.com/mike-ebert/2013/03/23/the-repository-pattern.html

**Comment by tansaku** [comment]
@johnnymo87 totally agree with you about a `app/repositories/*` location, however I am a little unsure about creating a name space like `Repositories::ProposedOrganisationEdit`  we don't have a similar namespace for services do we, i.e. there's no `Services::ProposedOrganisationEdit`

**Comment by johnnymo87** [comment]
We have `Queries::Organisations` [here](https://github.com/AgileVentures/LocalSupport/blob/proposed_org_repository_refactor/app/services/queries/organisations.rb).

**Comment by johnnymo87** [comment]
The `app/services` directory is wired up to be our default directory for business logic outside of rails, and any directories we nest in there (like `app/services/queries`) automagically work if namespaced properly. For example, if we had `app/services/presenters/organisations/show.rb`, it would need to be namespaced as `Presenters::Organisations::Show`. So in order to consistent, I should have suggested the `app/services/repositories` path to allow for `Repositories::ProposedOrganisationEdit`.

**Comment by tansaku** [comment]
@johnnymo87 it seems like any sub-dir of app gets wired up automatically.  What's the advantage of these name spaces?  

I guess there's a question of whether we consider a repository a subtype of service ...

**Comment by johnnymo87** [comment]
I believe other custom directories in app are *not* automatically wired up. I think the way we got the `app/services` to work is via [this line](https://github.com/AgileVentures/LocalSupport/blob/develop/config/application.rb#L21). Inside *subdirectories* of loaded directories, code gets auto-loaded via namespacing. For instance, if you type `Queries::Organisations` into the rails console, you'll see the constant is defined. But if you edit `app/services/queries/organisations.rb` and remove the outer `Queries` module (or change the spelling of it), and then `reload!` the rails console, you'll see that `Queries::Organisations` is not defined.

Yeah, I agree that `app/services` is a bit too specific. I'd like to rename it to `app/lib`.

**Comment by tansaku** [comment]
@johnnymo87 the thing that makes me think that other directories are auto-wired up is that in the current pull request the `app/repositories` directory is being pulled in and I even mis-spelt it the first time around.

Yeah, `app/lib` seems like a step in the right direction ... what about moving things into `lib`?  That's where we want to be ultimately right?  All the domain logic completely separate from the Rails app ...

In the first instance I'm tempted to merge this in as is, at least so I can start adding other CRUD methods to the repository.  The location can be moved easily in future.

Namespacing with modules seems like a good idea if we have a real explosion of types of things, but I worry that we're anticipating the complexity before it arrives as I think we've had a tendency to do in the past.  At the moment I feel like `ProposedOrganisationEditRepository` is at the edge of acceptability, and I'm not seeing a particular benefit from going with `Repository::ProposedOrganisationEdit`, but that might be a way to go once we've created a few more Repositories. 

I'm a big fan of watching the complexity emerge and then responding to it, rather than anticipating it when things might go in another direction.  I still need to get a better handle on the current services/queries set up ...

**Comment by johnnymo87** [comment]
Yeah, I'm just nitpicking, I think the code looks great. +1 merge it!

**Comment by mtc2013** [comment]
@tansaku this looks cool.

**Comment by tansaku** [comment]
@johnnymo87 @mtc2013 :+1:

**Pull request merged**